proposals/GlobalUsingDirective.md

```diff
@@ -2,13 +2,194 @@
 
 Syntax for a using directive is extended with an optional `global` keyword that can precede the `using` keyword:
 ```antlr
-using_directive
-  : 'global'? 'using' ('static' | name_equals)? name ';'
-  ;
+compilation_unit
+    : extern_alias_directive* global_using_directive* using_directive* global_attributes? namespace_member_declaration*
+    ;
+
+global_using_directive
+    : global_using_alias_directive
+    | global_using_namespace_directive
+    | global_using_static_directive
+    ;
+
+global_using_alias_directive
+    : 'global' 'using' identifier '=' namespace_or_type_name ';'
+    ;
+
+global_using_namespace_directive
+    : 'global' 'using' namespace_name ';'
+    ;
+    
+global_using_static_directive
+    : 'global' 'using' 'static' type_name ';'
+    ;
 ```
 
-- Global Using Directives are allowed only on the Compilation Unit level (cannot be used inside a namespace declaration).
-- Global Using directives, if any, must precede any non-global using directives. 
-- The scope of a Global Using Directive extends over the namespace member declarations and non-global using directives of all compilation units within the program.
-The scope of a Global Using Directive specifically does not include other Global Using Directives. Thus, peer Global Using Directives or those from a different
+- The *global_using_directive*s are allowed only on the Compilation Unit level (cannot be used inside a *namespace_declaration*).
+- The *global_using_directive*s, if any, must precede any *using_directive*s. 
+- The scope of a *global_using_directive*s extends over the namespace member declarations and *using_directive*s of all compilation units within the program.
+The scope of a *global_using_directive* specifically does not include other *global_using_directive*s. Thus, peer *global_using_directive*s or those from a different
 compilation unit do not affect each other, and the order in which they are written is insignificant.
+
+# Scopes 
+https://github.com/dotnet/csharplang/blob/master/spec/basic-concepts.md#scopes
+
+These are the relevant bullet points with proposed additions (which are **in bold**):
+*  The scope of name defined by an *extern_alias_directive* extends over the ***global_using_directive*s,** *using_directive*s, *global_attributes* and *namespace_member_declaration*s of its immediately containing compilation unit or namespace body. An *extern_alias_directive* does not contribute any new members to the underlying declaration space. In other words, an *extern_alias_directive* is not transitive, but, rather, affects only the compilation unit or namespace body in which it occurs.
```

Comment:
> The scope of name defined by an *extern_alias_directive* extends over the ***global_using_directive*s

How does this work? Since the extern directive only affects the compilation unit in which it appears, I'm not sure how this then applies when considering the global using directive from another compilation unit.

Reply:
> I'm not sure how this then applies when considering the global using directive from another compilation unit.

It doesn't. The statements says: "of its **immediately containing compilation unit** or namespace body."

---
In reply to: [589635783](https://github.com/dotnet/csharplang/pull/4499#discussion_r589635783) [](ancestors = 589635783)

Reply:
Then why even bother to list *global_using_directive* here at all? extern aliases can never apply to them.

---
In reply to: [589647367](https://github.com/dotnet/csharplang/pull/4499#discussion_r589647367) [](ancestors = 589647367,589635783)

Reply:
> Then why even bother to list global_using_directive here at all? extern aliases can never apply to them.

Because they do apply. Exactly as specified in that statement.

---
In reply to: [589650067](https://github.com/dotnet/csharplang/pull/4499#discussion_r589650067) [](ancestors = 589650067,589647367,589635783)

Reply:
> Because they do apply. Exactly as specified in that statement.

Then I'm back to my original question :). How does this work:

```cs
extern alias Other;
global using Other::A;
```

When a different compilation unit is looking at this global using alias, does it have to resolve the extern alias? If not, how does that global using work? And if so, then I'd rather have us introduce a global extern alias.

I don't necessarily expect you to answer this in this PR, but it would be good to have an open question about how extern aliases apply to global using directives.

---
In reply to: [589654069](https://github.com/dotnet/csharplang/pull/4499#discussion_r589654069) [](ancestors = 589654069,589650067,589647367,589635783)

Reply:
I think the statement with other added sections clearly describes how this works. "Other::A" should resolve to a namespace in context of the compilation unit where the using is "declared". Other compilation units never look at global aliases of each other. There is no need to do that.


---
In reply to: [589679320](https://github.com/dotnet/csharplang/pull/4499#discussion_r589679320) [](ancestors = 589679320,589654069,589650067,589647367,589635783)